pkg/runner/common.go

```diff
@@ -13,15 +22,81 @@ var (
 	controlGateway = "192.18.0.1"
 )
 
-func nextDataNetwork(lenNetworks int) (string, string, error) {
+func nextDataNetwork(lenNetworks int) (*net.IPNet, string, error) {
 	if lenNetworks > 4095 {
-		return "", "", errors.New("space exhausted")
+		return nil, "", errors.New("space exhausted")
 	}
 	a := 16 + lenNetworks/256
 	b := 0 + lenNetworks%256
 
-	subnet := fmt.Sprintf("%d.%d.0.0/16", a, b)
-	gateway := fmt.Sprintf("%d.%d.0.1", a, b)
+	sn := fmt.Sprintf("%d.%d.0.0/16", a, b)
+	gw := fmt.Sprintf("%d.%d.0.1", a, b)
+
+	_, subnet, err := net.ParseCIDR(sn)
+	return subnet, gw, err
+}
+
+func zipRunOutputs(ctx context.Context, basedir string, input *api.CollectionInput, w io.Writer) error {
+	pattern := filepath.Join(basedir, "*", input.RunID)
+
+	matches, err := filepath.Glob(pattern)
+	if err != nil {
+		return err
+	}
+
+	if len(matches) != 1 {
+		return fmt.Errorf("run ID %s not found with runner %s", input.RunID, input.RunnerID)
+	}
+
+	dir := matches[0]
+
+	if fi, err := os.Stat(dir); err != nil {
+		return err
+	} else if !fi.IsDir() {
+		return fmt.Errorf("internal error: not a directory when accessing run outputs")
+	}
+
+	wz := zip.NewWriter(w)
+	defer wz.Close()
+	defer wz.Flush()
+
+	base := filepath.Base(dir)
+	return filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		header, err := zip.FileInfoHeader(info)
```

Comment:
I think this part can be simplified - just call `zip.Create` which does the `FileHeader` for you and returns a writer.

But if this works, it works :)